sycl-fusion/passes/target/TargetFusionInfo.h

```diff
@@ -0,0 +1,186 @@
+
+  virtual void notifyFunctionsDelete(llvm::ArrayRef<Function *> Funcs) const {
+    (void)Funcs;
+  }
+
+  virtual void addKernelFunction(Function *KernelFunc) const {
+    (void)KernelFunc;
+  }
+
+  virtual void postProcessKernel(Function *KernelFunc) const {
+    (void)KernelFunc;
+  }
+
+  virtual ArrayRef<StringRef> getKernelMetadataKeys() const { return {}; }
+
+  virtual void createBarrierCall(IRBuilderBase &Builder,
+                                 int BarrierFlags) const = 0;
+
+  virtual unsigned getPrivateAddressSpace() const = 0;
+
+  virtual unsigned getLocalAddressSpace() const = 0;
+
+  virtual void updateAddressSpaceMetadata(Function *KernelFunc,
+                                          ArrayRef<size_t> LocalSize,
+                                          unsigned AddressSpace) const {
+    (void)KernelFunc;
+    (void)LocalSize;
+  }
+
+protected:
+  explicit TargetFusionInfoImpl(llvm::Module *Mod) : LLVMMod{Mod} {};
+
+  llvm::Module *LLVMMod;
+
+  friend class TargetFusionInfo;
+};
+
+class SPIRVTargetFusionInfo : public TargetFusionInfoImpl {
+public:
+  void addKernelFunction(Function *KernelFunc) const override;
+
+  ArrayRef<StringRef> getKernelMetadataKeys() const override;
+
+  void postProcessKernel(Function *KernelFunc) const override;
+
+  void createBarrierCall(IRBuilderBase &Builder,
+                         int BarrierFlags) const override;
+
+  // Corresponds to definition of spir_private and spir_local in
+  // "clang/lib/Basic/Target/SPIR.h", "SPIRDefIsGenMap".
+  unsigned getPrivateAddressSpace() const override { return 0; }
+  unsigned getLocalAddressSpace() const override { return 3; }
+
+  void updateAddressSpaceMetadata(Function *KernelFunc,
+                                  ArrayRef<size_t> LocalSize,
+                                  unsigned AddressSpace) const override;
+
+private:
+  using TargetFusionInfoImpl::TargetFusionInfoImpl;
+};
+
+class NVPTXTargetFusionInfo : public TargetFusionInfoImpl {
+public:
+  void notifyFunctionsDelete(llvm::ArrayRef<Function *> Funcs) const override;
+
+  void addKernelFunction(Function *KernelFunc) const override;
+
+  ArrayRef<StringRef> getKernelMetadataKeys() const override;
+
+  void createBarrierCall(IRBuilderBase &Builder,
+                         int BarrierFlags) const override;
+
+  // Corresponds to the definitions in the LLVM NVPTX backend user guide:
+  // https://llvm.org/docs/NVPTXUsage.html#address-spaces
+  unsigned getPrivateAddressSpace() const override { return 0; }
+  unsigned getLocalAddressSpace() const override { return 3; }
+
+private:
+  using TargetFusionInfoImpl::TargetFusionInfoImpl;
+};
+
+///
+/// Common interface to target-specific logic around handling of kernel
+/// functions.
+class TargetFusionInfo {
+public:
+  ///
+  /// Create the correct target-specific implementation based on the target
+  /// triple of \p Module.
+  static TargetFusionInfo getTargetFusionInfo(llvm::Module *Module);
```

Comment:
```suggestion
  static TargetFusionInfo getTargetFusionInfo(llvm::Module *Module);
```
NIT: Maybe passing triple straight into here would make sense?